cmd/mount_zfs/mount_zfs.c

```diff
@@ -47,63 +47,44 @@ libzfs_handle_t *g_zfs;
  * is expected to be of the form pool/dataset, however may also refer to
  * a block device if that device contains a valid zfs label.
  */
-static char *
-parse_dataset(char *dataset)
+static void
+parse_dataset(char (*dataset)[PATH_MAX])
 {
-	char cwd[PATH_MAX];
-	struct stat64 statbuf;
-	int error;
-	int len;
-
 	/*
 	 * We expect a pool/dataset to be provided, however if we're
 	 * given a device which is a member of a zpool we attempt to
 	 * extract the pool name stored in the label.  Given the pool
 	 * name we can mount the root dataset.
 	 */
-	error = stat64(dataset, &statbuf);
-	if (error == 0) {
+	int fd = open((const char *)dataset, O_RDONLY);
+	if (fd > 0) {
 		nvlist_t *config;
-		char *name;
-		int fd;
-
-		fd = open(dataset, O_RDONLY);
-		if (fd < 0)
-			goto out;
-
-		error = zpool_read_label(fd, &config, NULL);
+		if (zpool_read_label(fd, &config, NULL) != 0)
+			config = NULL;
 		(void) close(fd);
-		if (error)
-			goto out;
-
-		error = nvlist_lookup_string(config,
-		    ZPOOL_CONFIG_POOL_NAME, &name);
-		if (error) {
-			nvlist_free(config);
-		} else {
-			dataset = strdup(name);
-			nvlist_free(config);
-			return (dataset);
-		}
+
+		char *name = NULL;
+		nvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME, &name);
+		if (name)
+			(void) strlcpy(*dataset, name, PATH_MAX);
+		nvlist_free(config);
+		return;
 	}
-out:
+
 	/*
 	 * If a file or directory in your current working directory is
 	 * named 'dataset' then mount(8) will prepend your current working
 	 * directory to the dataset.  There is no way to prevent this
 	 * behavior so we simply check for it and strip the prepended
 	 * patch when it is added.
 	 */
-	if (getcwd(cwd, PATH_MAX) == NULL)
-		return (dataset);
-
-	len = strlen(cwd);
-
-	/* Do not add one when cwd already ends in a trailing '/' */
-	if (strncmp(cwd, dataset, len) == 0)
-		return (dataset + len + (cwd[len-1] != '/'));
-
-	return (dataset);
+	char cwd[PATH_MAX];
+	if (getcwd(cwd, PATH_MAX) != NULL) {
+		int len = strlen(cwd);
+		/* Do not add one when cwd already ends in a trailing '/' */
+		if (strncmp(cwd, *dataset, len) == 0)
+			(void) strlcpy(*dataset, cwd, len+(cwd[len-1] != '/'));
+	}
```

Comment:
I'm rather puzzled by this bit. The path can only end with '/' in one case: where it literally is '/', which also seems to be recognized in the original commit which introduced this part ( a6cba65ccae5a5458f9ab4ac11020716d30ea7f7 ). iow, you can drop strlen and only check for `cwd[0] == '/' && cwd[0] == '\0'`

Reply:
Yeah, the issue of trailing slashes from `getcwd` was addressed with POSIX.1-2008₁, but more importantly: my patch breaks the original mount(8) `$PWD` prefix bug workaround. I'm writing some test cases to detect this - before my next revision. This workaround isn't really needed when path canonicalization is disabled, which is the ideal solution, but outside of the scope of this patch. I'll get to that one next.
₁[Austin Group Interpretation 1003.1-2001 #140](https://collaboration.opengroup.org/austin/interps/documents/14363/AI-140.txt)